doc/developer/design/20230531_system_privileges.md

```diff
@@ -0,0 +1,97 @@
+keywords. For example, `ALTER ROLE r CREATEDB` will grant the ability to create databases to
+role `r`. `ALTER ROLE r CREATECLUSTER` will grant the ability to create clusters to role `r`. Role
+attributes cannot be inherited through role membership.
+
+Role attributes always exist in a boolean state. To change the state of a role attribute you must
+alter the role. For example `ALTER ROLE joe CREATEDB` sets the `CREATEDB` attribute to `true`,
+and `ALTER ROLE NOCREATEDB` sets the `CREATEDB` attribute to `false`.
+
+On the other hand, privileges either exist or they don't exist. You create them by granting a
+privilege via `GRANT ...` and you destroy them via `REVOKE ...`.
+
+Role attributes present some problems as they currently exist:
+
+- It can be confusing that there are two separate privilege primitives that control a role's
+  privilege, and they have different semantics.
+- Role attributes are not inherited which makes it cumbersome to grant attributes to a large
+  number of roles.
+- Users are not able to easily grant the ability to create top level objects to many current and
+  future users in a short amount of time.
+
+## Goals
+
+- Reduce confusion around privileges.
+- Allow the ability to create databases, clusters, and roles to be inherited through role
+  membership.
+- Allow users to easily grant the ability to create top level objects to many current and future
+  users in a short amount of time.
+- Maintain PostgreSQL compatibility.
+
+## Overview
+
+We will add a new type of privilege called System Privileges. They will look and act just like
+normal object privileges, but they will grant the abilities that would otherwise be granted through
+role attributes. They will exist in parallel with role attributes.
+
+NOTE: Other databases, such as CockroachDB and Snowflake, have some form of system privileges that
+control similar privileges to role attributes.
+
+## Detailed description
+
+System privileges will be granted with the following syntax:
+`GRANT <privilege-specification> ON SYSTEM TO <role-specification>`
+
+System privileges will be revoked with the following syntax:
+`REVOKE <privilege-specification> ON SYSTEM FROM <role-specification>`
+
+System privileges and role attributes will interact in the following ways:
+
+- Granting a system privilege will grant the privilege AND set the attribute to true.
+- Revoking a system privilege will revoke the privilege AND set the attribute to false.
+- Altering a role attribute will affect the role attribute but will not change the system privilege.
+- When checking to see if a role is allowed to execute a command we will check if they have the
+  system privilege or the role attribute.
+
+This approach has the following nice properties:
+
+- The ability to use system privileges without thinking about role attributes.
+- The ability to use role attributes without thinking about system privileges.
+- System privileges always take priority over role attributes when using both.
```

Comment:
These two concepts seem to conflict with one another? If I set `NOCREATEDB` for a given role, that role might still have the ability to create a database via it's system privileges?

Reply:
> If I set `NOCREATEDB` for a given role, that role might still have the ability to create a database via it's system privileges?

Yes that's correct. I think this is the roughest edge of the entire design, though I haven't been able to come up with a better solution if we want both role attributes and system privileges to exist concurrently. 

On the bright side, this somewhat matches existing semantics for privileges. If I type `REVOKE SELECT on t FROM r`, then `r` may still have the ability to select from `t` if it inherits that privilege from another role.